File: src/main/java/frc/robot/subsystems/intake/IntakeIO.java
package frc.robot.subsystems.intake;

import org.littletonrobotics.junction.AutoLog;

import edu.wpi.first.math.geometry.Rotation2d;

public interface IntakeIO {

    @AutoLog
    public static class IntakeIOInputs {

        /**
         * Velocity of intake arm in rotations/second
         */
        public double armVelocity = 0.0;

        /**
         * Voltage of intake arm motor in volts
         */
        public double armAppliedVoltage = 0.0;

        /**
         * Current of intake arm motor in amperes
         */
        public double armCurrentAmps = 0.0;

        /**
         * Temperature of intake arm motor in celcius
         */
        public double armTempCelcius = 0.0;

        /**
         * Position of intake arm in rotations
         */
        public Rotation2d armPosition = Rotation2d.kZero;

        /**
         * Velocity of roller motor in rotations/second
         */
        public double rollerVelocity = 0.0;

        /**
         * Voltage of roller motor in volts
         */
        public double rollerAppliedVoltage = 0.0;

        /**
         * Current of roller motor in amperes
         */
        public double rollerCurrentAmps = 0.0;

        /**
         * Temperature of roller motor in celcius
         */
        public double rollerTempCelcius = 0.0;

        public boolean isDeployed = false;
    }

    public default void updateInputs(IntakeIOInputs inputs) {}

    public default void stopArmMotor() {}

    public default void stopRollerMotor() {}

    public default void setArmVoltage(double volts) {}

    public default void setRollerVoltage(double volts) {}

    public default void setArmPosition(Rotation2d position) {}

    public default void setRollerSpeed(double speed) {}

    public default void setArmIdleMode(boolean isBrake) {}

    public default void setRollerIdleMode(boolean isBrake) {}

    public default void resetEncoder(Rotation2d position) {}

    public default void periodic() {};

}

File: src/main/java/frc/robot/subsystems/intake/IntakeIO.kt
package frc.robot.subsystems.intake

import com.lobstahbots.junction.AutoLogKt
import edu.wpi.first.math.geometry.Rotation2d
import com.lobstahbots.units.*

interface IntakeIO {
    @AutoLogKt
    open class IntakeIOInputs {
        /**
         * Velocity of intake arm in rotations/second
         */
        var armVelocity = 0.rotationsPerSecond

        /**
         * Voltage of intake arm motor in volts
         */
        var armAppliedVoltage = 0.volts

        /**
         * Current of intake arm motor in amperes
         */
        var armCurrent = 0.amps

        /**
         * Temperature of intake arm motor in celcius
         */
        var armTemp = 0.celsius

        /**
         * Position of intake arm in rotations
         */
        var armPosition: Rotation2d = Rotation2d.kZero

        /**
         * Velocity of roller motor in rotations/second
         */
        var rollerVelocity = 0.rotationsPerSecond

        /**
         * Voltage of roller motor in volts
         */
        var rollerAppliedVoltage = 0.volts

        /**
         * Current of roller motor in amperes
         */
        var rollerCurrent = 0.amps

        /**
         * Temperature of roller motor in Celsius
         */
        var rollerTemp = 0.celsius

        var isDeployed: Boolean = false
    }

    fun updateInputs(inputs: IntakeIOInputs) {}

    fun stopArmMotor() {}

    fun stopRollerMotor() {}

    fun setArmVoltage(volts: Double) {}

    fun setRollerVoltage(volts: Double) {}

    fun setArmPosition(position: Rotation2d) {}

    fun setRollerSpeed(speed: Double) {}

    fun setArmIdleMode(isBrake: Boolean) {}

    fun setRollerIdleMode(isBrake: Boolean) {}

    fun resetEncoder(position: Rotation2d) {}

    fun periodic() {}
}
